File: Top 100 Codes/07-Codes related to Matrix/08-Kth smallest element in a row-column wise sorted matrix/Kthsmallestelementinarowcolumnwisesortedmatrix.py
def count_less_equal(matrix, mid, n):
    count = 0
    row = n - 1
    col = 0

    while row >= 0 and col < n:
        if matrix[row][col] <= mid:
            count += row + 1
            col += 1
        else:
            row -= 1
    return count

def kth_smallest(matrix, k):
    n = len(matrix)
    low = matrix[0][0]
    high = matrix[-1][-1]

    while low < high:
        mid = (low + high) // 2
        count = count_less_equal(matrix, mid, n)

        if count < k:
            low = mid + 1
        else:
            high = mid
    return low

# Test
matrix = [
    [1, 5, 9],
    [10, 11, 13],
    [12, 13, 15]
]
k = 8
print("Kth Smallest Element is:", kth_smallest(matrix, k))

File: Top 100 Codes/07-Codes related to Matrix/08-Kth smallest element in a row-column wise sorted matrix/Kthsmallestelementinarowcolumnwisesortedmatrix.cpp
#include <iostream>
#include <vector>
using namespace std;

int countLessEqual(vector<vector<int>>& matrix, int mid, int n) {
    int count = 0;
    int row = n - 1, col = 0;

    while (row >= 0 && col < n) {
        if (matrix[row][col] <= mid) {
            count += row + 1;
            col++;
        } else {
            row--;
        }
    }
    return count;
}

int kthSmallest(vector<vector<int>>& matrix, int k) {
    int n = matrix.size();
    int low = matrix[0][0];
    int high = matrix[n - 1][n - 1];

    while (low < high) {
        int mid = low + (high - low) / 2;
        int count = countLessEqual(matrix, mid, n);

        if (count < k)
            low = mid + 1;
        else
            high = mid;
    }
    return low;
}

int main() {
    vector<vector<int>> matrix = {
        {1, 5, 9},
        {10, 11, 13},
        {12, 13, 15}
    };
    int k = 8;
    cout << "Kth Smallest Element is: " << kthSmallest(matrix, k) << endl;
    return 0;
}
